enhancements/network/vrf-cni.md

```diff
@@ -0,0 +1,159 @@
+
+In this way, the network interface returned by any CNI can be isolated in its own VRF, satisfying the above requirements without the need of additional capabilities.
+
+### User Stories [optional]
+
+#### Story 1
+
+As a user, I want to be able to optionally specify a VRF which the secondary nic is attached to.
+
+### Implementation Details/Notes/Constraints
+
+By adding several networks with the same `vrf` name, the CNI will ensure that they are all added to the same vrf, creating it only the first time.
+
+A `NetworkAttachmentDefinition` for MacVlan would look like:
+
+```yaml
+apiVersion: "k8s.cni.cncf.io/v1"
+kind: NetworkAttachmentDefinition
+metadata:
+  name: macvlan-vrf-conf
+spec:
+  config: '{
+              "cniVersion": "0.3.1",
+              "plugins": [
+                {
+                  "type": "macvlan",
+                  "mode": "bridge",
+                  "ipam": {
+                    "type": "host-local",
+                    "subnet": "192.168.1.0/24",
+                    "rangeStart": "192.168.1.200",
+                    "rangeEnd": "192.168.1.216",
+                  }
+                },
+                {
+                  "type": "vrf",
+                  "vrfname": "blue"
+                }
+              ]
+            }'
+```
+
+The same example for a SR-IOV device would look like:
+
+```yaml
+apiVersion: "k8s.cni.cncf.io/v1"
+kind: NetworkAttachmentDefinition
+metadata:
+  name: sriov-vrf-conf
+  annotations:
+    k8s.v1.cni.cncf.io/resourceName: openshift.io/resource_id
+spec:
+  config: '{
+              "cniVersion": "0.3.1",
+              "plugins": [
+                {
+                  "type": "sriov",
+                  "ipam": {
+                    "type": "host-local",
+                    "subnet": "192.168.1.0/24",
+                    "rangeStart": "192.168.1.200",
+                    "rangeEnd": "192.168.1.216",
+                  }
+                },
+                {
+                  "type": "vrf",
+                  "vrfname": "blue"
+                }
+              ]
+            }'
+```
+
+This can be supported at `NetworkAttachmentDefinition` level, or by documenting (and validating) chained CNIs in the `Network` CRD of `operator.openshift.io/v1`.
+
+At the same time, the `SriovNetwork` type from the SR-IOV operator needs to be extended for chaining meta CNIs.
+
+The CNIs we intend to support are SR-IOV and MacVlan.
+
+### Risks and Mitigations
+
+## Design Details
+
+This would be shipped as a new CNI plugin, ideally being part of the container networking plugins [repo](https://github.com/containernetworking/plugins),
+which is then consumed in our [openshift variant](https://github.com/openshift/containernetworking-plugins). If not feasible, other alternatives
+may be considered for shipping it.
+
+The CNI implementation is pretty simple: it takes the result of the main CNI, creates the VRF in the pod's namespace and assignes the interface to the VRF.
```

Comment:
what kernel versions are we addressing, the docs https://www.kernel.org/doc/Documentation/networking/vrf.txt mentions

> Using iproute2 for VRFs
=======================
iproute2 supports the vrf keyword as of v4.7. For backwards compatibility this
section lists both commands where appropriate -- with the vrf keyword and the
older form without it.

Reply:
I think the 4.7 referred there is the iproute2 version, not the kernel version. We are targeting rhel 8.3(+) which I think ships a more recent version. I remember trying to manually create a vrf in a pod and it worked in 4.6.